src/consul.rs

```diff
@@ -0,0 +1,399 @@
+        }
+    }
+
+    async fn wait_for_lock(&self) -> Result<(), anyhow::Error> {
+        let mut consul_index: Option<u64> = None;
+        loop {
+            let lock_url = self.kv_api_base_url.join(CONSUL_STORE_KEY)?;
+
+            let mut req = self.http_client.get(lock_url);
+
+            // Set dc if it is provided in the config
+            if let Some(dc) = &self.datacenter {
+                req = req.query(&[("dc", dc)]);
+            }
+
+            if let Some(index) = consul_index.take() {
+                req = req.query(&[("index", &index.to_string())]);
+            }
+
+            let response = req.send().await?;
+
+            consul_index = response
+                .headers()
+                .get("X-Consul-Index")
+                .and_then(|value| value.to_str().ok())
+                .and_then(|value| value.parse().ok());
+
+            let kvs = response.json::<Vec<ConsulKVResponse>>().await?;
+
+            for kv in kvs {
+                if kv.key == CONSUL_STORE_KEY && kv.session.is_none() {
+                    println!("=====> lock is free, returning");
+                    return Ok(());
+                }
+            }
+        }
+    }
+
+    /// Retrieves a list of all registered services and parses their tags into DnsTag
+    pub async fn fetch_service_tags(
+        &self,
+        consul_index: &mut Option<String>,
+    ) -> Result<Vec<DnsRecord>, anyhow::Error> {
+        let services_url = self.catalog_api_base_url.join("services")?;
+
+        let mut req = self.http_client.get(services_url);
+
+        // Set dc if it is provided in the config
+        if let Some(dc) = &self.datacenter {
+            req = req.query(&[("dc", dc)]);
+        }
+
+        if let Some(index) = consul_index {
+            req = req.query(&[("index", &index.to_string())]);
+        }
+
+        // Add a filter to only match "normal" Consul services
+        req = req.query(&[(
+            "filter",
+            r#"ServiceKind == "" and ServiceTags contains "external-dns.enable=true""#,
+        )]);
+
+        let response = req.send().await?.error_for_status()?;
+
+        if let Some(index_header) = response.headers().get("X-Consul-Index") {
+            if let Ok(index_str) = index_header.to_str() {
+                let _ = consul_index.insert(index_str.to_string());
+            } else {
+                eprintln!("Failed to convert HeaderValue to string");
+            }
+        }
+
+        let records = response.json::<HashMap<String, Vec<String>>>().await?;
+
+        let dns_tags = records
+            .into_iter()
+            .flat_map(|(_service_name, tags)| parse_dns_tags(tags))
+            .collect();
+
+        Ok(dns_tags)
+    }
+
+    /// Stores a single DNS record in Consul.
+    /// This function fetches the current state of DNS records, updates it with the new record,
+    /// and then re-stores the updated state back into Consul.
+    pub async fn store_dns_record(
```

Comment:
This doc comment is wrong now, but it feels like the method name is also rather misleading.  I wonder whether something like `stage_dns_record` would make it more obvious that this just stores it locally and it's only when we call `store_all_dns_records` that they're actually sent to Consul.

On the other hand, this API overall feels a bit awkward.  Maybe it would actually make more sense for the hashmap to instead be stored as a local variable in `process_dns_records`, and then passed by reference to `store_all_dns_records`, since it's semantically rather weird for the `ConsulClient` to store such local-only state itself.

Reply:
used local variable in `process_dns_records`